### ISSUE TITLE
`no-implicit-side-effects` - disallow `ExpressionStatement`s

### ISSUE THREAD

**Issue by jussi-kalliokoski** [eslint/eslint#4332]
When programmers new to functional concepts are introduced to codebases using React and/or Redux, they often don't quite understand what "having no side effects" or "not mutating objects" mean. Hence it's common to see them make changes to otherwise pure functions that mutate an object or make a function call without using the value, causing bugs that are hard to debug.

The proposed rule would disallow all nodes of type `ExpressionStatement`:

```javascript
// Not OK:
a = 1;
b.c = 2;
foo();
bar.qoo();
z.y++;
[1, 2, 3].forEach(n => console.log(n));
(function (){}());
// etc...
```

while things like

```
var h = 3;
let i = 4;
const j = bar();
return foo();
```
remain valid.

As an optional "I know what I'm doing" escape hatch, there could be an option to allow `ExpressionStatement`s when they're prefixed with `void` to make side effects explicit:

```javascript
// valid
void (a = 1);
void (b.c = 2);
void foo();
void bar.qoo();
void z.y++;
void [1, 2, 3].forEach(n => console.log(n));
void function (){}();
```

The rule is geared towards functional programming, but can be used with OOP as well, although with most OOP tools you would end up with a lot of `void` prefixes. However with a reasonably idiomatic Redux + React codebase, you only end up with a few `void`s (most notably for `ReactDOM.render()`).

There are some pitfalls that this rule would not touch, for example short body arrow functions can conceal side effects because they appear as return values, but same is true of putting the side effect in any place where expressions are accepted. Using this rule with

- `no-class-assign`
- `no-cond-assign`
- `no-const-assign`
- `no-func-assign`
- `no-new`
- `no-param-reassign`
- `no-plusplus`
- `no-return-assign`
- `no-sequences`
- `operator-assignment` (`"never"`)
- `prefer-const`

covers most of the common cases and requires the programmer to take conscious effort to actually produce a side effect without making it explicit.

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by jussi-kalliokoski** [comment]
Oh, and FWIW I have a potential implementation and a few tests for this rule. I just think this sort of rule belongs in the core given the recent rise in functional paradigms with JS, so I thought I'd see if people agree before I publish it as a plugin instead.

**Comment by michaelficarra** [comment]
This would be essentially the same as no-with but for a different type of node.

**Comment by gyandeeps** [comment]
There is an exiting rule which helps you to disallow certain types of nodes: http://eslint.org/docs/rules/no-restricted-syntax

**Comment by jussi-kalliokoski** [comment]
> I am not sure if this fully fits your scenario.

Sans the `void` escape hatch and a good description of why this specific thing is an error.

**Comment by michaelficarra** [comment]
@jussi-kalliokoski You would probably also want to disallow `void`, as you mentioned, and sequence expressions (`a, b`).

**Comment by ilyavolodin** [comment]
`void` as an escape hatch seems pretty arbitrary to me. I actually think that comments that disable rule for a line is more descriptive. In general, it seems we can already cover this scenario with existing rule, and creating another rule just for clearer error message will conflict with existing rules.

**Comment by michaelficarra** [comment]
This rule's intent, as I understand it, is to catch all occurrences of the result of an expression being unused. In my opinion, this should include void. And sequences.

**Comment by ilyavolodin** [comment]
@michaelficarra `no-restricted-syntax` can take more than one type of expression. So you could disable `ExpressionStatement` as well as `SequenceExpression` (which is always nested under `ExpressionStatement` as far as I can tell, correct me if I'm wrong here). `void` would be a problem, however, agree.

**Comment by jussi-kalliokoski** [comment]
> You would probably also want to disallow void

No, I want to use void. x)

> sequence expressions (a, b)

As I mentioned, this rule would go well with `no-sequences`.

> void as an escape hatch seems pretty arbitrary to me.

It's not completely arbitrary though. It's an explicit marker that "the value of this expression is not used", that's the only thing that it does, makes any value unusable. It's also used in other languages to mark functions that return no value, so it's pretty fitting.

**Comment by michaelficarra** [comment]
@ilyavolodin Why do you think sequences are only ever in statement position? They can be used anywhere an expression is expected.

**Comment by nzakas** [comment]
Please use `no-restricted-syntax` for this. We aren't going to add another rule that just flags specific node types.

**Comment by jussi-kalliokoski** [comment]
Fair enough, but `no-restricted-syntax` doesn't cover the intended use (allowing them when prefixed with `void`) - making `void` optional was just an idea for configurability. However, `no-restricted-syntax` does makes that configurability redundant, so I dropped that from my implementation and published as a plugin instead: [jussi-kalliokoski/eslint-plugin-no-implicit-side-effects](https://github.com/jussi-kalliokoski/eslint-plugin-no-implicit-side-effects).
